File: packages/js/node/src/swap/fixedfloat-currencies.ts
/**
 * FixedFloat `/ccies` discovery: which provider currency codes stand for
 * OpenReceive's pay-in assets and for the BTC Lightning payout, the cache
 * (de)serialization of that resolution, and the rate-pair keys it implies.
 * `/ccies` carries availability and display metadata only — never limits.
 */

import { recordOrEmpty } from "@openreceive/core";
import {
  getOpenReceiveSwapAssetInfo,
  isOpenReceiveLightningNetwork,
  listOpenReceiveSwapAssetInfo,
  openReceiveSwapNetworkMatches,
  type SwapPayInAsset,
} from "./assets.ts";
import { optionalStringField } from "./fixedfloat-fields.ts";
import { fixedFloatRatesPairKey } from "./fixedfloat-rates.ts";

export interface FixedFloatCurrency {
  readonly code: string;
  readonly coin: string;
  readonly network: string;
  readonly recv?: boolean;
  readonly send?: boolean;
}

export interface FixedFloatCurrencyResolution {
  readonly fetched_at: number;
  readonly pay_in: ReadonlyMap<SwapPayInAsset, FixedFloatCurrency>;
  readonly lightning: FixedFloatCurrency;
}

/**
 * Match a `/ccies` body against OpenReceive's pay-in asset list and pick the
 * Lightning payout currency: `lightningCcy` when configured, otherwise the
 * first sendable BTC Lightning entry.
 */
export function resolveFixedFloatCurrencies(
  data: unknown,
  input: { readonly fetchedAt: number; readonly lightningCcy: string | undefined },
): FixedFloatCurrencyResolution {
  const currencies = readFixedFloatCurrencies(data);
  const payIn = new Map<SwapPayInAsset, FixedFloatCurrency>();
  for (const asset of listOpenReceiveSwapAssetInfo()) {
    const found = currencies.find(
      (currency) =>
        currency.coin.toUpperCase() === asset.coin &&
        openReceiveSwapNetworkMatches(asset.network, currency.network) &&
        // /ccies recv=false means FixedFloat will not accept deposits for this
        // currency — omit it from the catalog rather than failing at /create.
        currency.recv !== false,
    );
    if (found !== undefined) payIn.set(asset.pay_in_asset, found);
  }

  const lightningCurrency =
    input.lightningCcy === undefined
      ? currencies.find(
          (currency) =>
            currency.coin.toUpperCase() === "BTC" &&
            isOpenReceiveLightningNetwork(currency.network) &&
            // Payout side must be sendable to the merchant's bolt11.
            currency.send !== false,
        )
      : currencies.find(
          (currency) => currency.code === input.lightningCcy && currency.send !== false,
        );
  if (lightningCurrency === undefined) {
    throw new Error("FixedFloat /ccies did not include a BTC Lightning payout currency.");
  }

  return {
    fetched_at: input.fetchedAt,
    pay_in: payIn,
    lightning: lightningCurrency,
  };
}

export function serializeFixedFloatCurrencyResolution(
  resolution: FixedFloatCurrencyResolution,
): string {
  return JSON.stringify({
    fetched_at: resolution.fetched_at,
    pay_in: Array.from(resolution.pay_in.entries()),
    lightning: resolution.lightning,
  });
}

export function deserializeFixedFloatCurrencyResolution(
  value: string,
): FixedFloatCurrencyResolution {
  const parsed = JSON.parse(value) as {
    readonly fetched_at: number;
    readonly pay_in: readonly (readonly [SwapPayInAsset, FixedFloatCurrency])[];
    readonly lightning: FixedFloatCurrency;
  };
  return {
    fetched_at: parsed.fetched_at,
    pay_in: new Map(parsed.pay_in),
    lightning: parsed.lightning,
  };
}

/** The provider code for a pay-in asset, or the payer-facing "not supported" error. */
export function requiredFixedFloatCurrency(
  resolution: FixedFloatCurrencyResolution,
  payInAsset: SwapPayInAsset,
): string {
  const currency = resolution.pay_in.get(payInAsset);
  if (currency === undefined) {
    const label = getOpenReceiveSwapAssetInfo(payInAsset).pay_in_asset;
    throw new Error(`FixedFloat does not currently support ${label}.`);
  }
  return currency.code;
}

export function openReceiveFixedFloatRatePairKeys(
  resolution: FixedFloatCurrencyResolution,
): Set<string> {
  const keys = new Set<string>();
  for (const currency of resolution.pay_in.values()) {
    keys.add(fixedFloatRatesPairKey(currency.code, resolution.lightning.code));
  }
  return keys;
}

function readFixedFloatCurrencies(data: unknown): FixedFloatCurrency[] {
  const record = recordOrEmpty(data);
  const items = Array.isArray(data)
    ? data
    : Array.isArray(record.ccies)
      ? record.ccies
      : Array.isArray(record.currencies)
        ? record.currencies
        : [];
  const currencies: FixedFloatCurrency[] = [];
  for (const item of items) {
    const record = recordOrEmpty(item);
    const code = optionalStringField(record, "code") ?? optionalStringField(record, "ticker");
    const coin =
      optionalStringField(record, "coin") ??
      optionalStringField(record, "currency") ??
      optionalStringField(record, "symbol");
    const network =
      optionalStringField(record, "network") ??
      optionalStringField(record, "chain") ??
      optionalStringField(record, "networkName") ??
      optionalStringField(record, "name");
    if (code !== undefined && coin !== undefined && network !== undefined) {
      currencies.push({
        code,
        coin: coin.toUpperCase(),
        network,
        ...(typeof record.recv === "boolean" ? { recv: record.recv } : {}),
        ...(typeof record.send === "boolean" ? { send: record.send } : {}),
      });
    }
  }
  return currencies;
}

File: packages/dotnet/BTCPayServer.Plugins.OpenReceive/Swaps/FixedFloatCurrencies.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Nodes;

namespace BTCPayServer.Plugins.OpenReceive.Swaps;

/// <summary>One entry of a FixedFloat <c>/ccies</c> body.</summary>
public sealed record FixedFloatCurrency(string Code, string Coin, string Network, bool? Recv = null, bool? Send = null);

/// <summary>
/// Which provider currency codes stand for OpenReceive's pay-in assets (keyed by
/// pay_in_asset) and for the BTC Lightning payout.
/// </summary>
public sealed record FixedFloatCurrencyResolution(
    long FetchedAt,
    IReadOnlyDictionary<string, FixedFloatCurrency> PayIn,
    FixedFloatCurrency Lightning);

/// <summary>
/// FixedFloat <c>/ccies</c> discovery. <c>/ccies</c> carries availability and display
/// metadata only — never limits.
/// </summary>
public static class FixedFloatCurrencies
{
    /// <summary>
    /// Match a <c>/ccies</c> body against OpenReceive's pay-in asset list and pick the
    /// Lightning payout currency: <paramref name="lightningCcy"/> when configured,
    /// otherwise the first sendable BTC Lightning entry.
    /// </summary>
    public static FixedFloatCurrencyResolution Resolve(JsonNode? data, long fetchedAt, string? lightningCcy)
    {
        var currencies = ReadCurrencies(data);
        var payIn = new Dictionary<string, FixedFloatCurrency>(StringComparer.Ordinal);
        foreach (var asset in SwapAssets.List())
        {
            var found = currencies.FirstOrDefault(currency =>
                currency.Coin.ToUpperInvariant() == asset.Coin &&
                SwapAssets.SwapNetworkMatches(asset.Network, currency.Network) &&
                // /ccies recv=false means FixedFloat will not accept deposits for this
                // currency — omit it from the catalog rather than failing at /create.
                currency.Recv != false);
            if (found is not null) payIn[asset.PayInAsset] = found;
        }

        var lightning = lightningCcy is null
            ? currencies.FirstOrDefault(currency =>
                currency.Coin.ToUpperInvariant() == "BTC" &&
                SwapAssets.IsLightningNetwork(currency.Network) &&
                // Payout side must be sendable to the merchant's bolt11.
                currency.Send != false)
            : currencies.FirstOrDefault(currency => currency.Code == lightningCcy && currency.Send != false);
        if (lightning is null)
        {
            throw new InvalidOperationException("FixedFloat /ccies did not include a BTC Lightning payout currency.");
        }

        return new FixedFloatCurrencyResolution(fetchedAt, payIn, lightning);
    }

    /// <summary>The provider code for a pay-in asset, or the payer-facing "not supported" error.</summary>
    public static string RequiredCurrency(FixedFloatCurrencyResolution resolution, string payInAsset)
    {
        if (resolution.PayIn.TryGetValue(payInAsset, out var currency)) return currency.Code;
        var label = SwapAssets.GetInfo(payInAsset).PayInAsset;
        throw new InvalidOperationException($"FixedFloat does not currently support {label}.");
    }

    public static IReadOnlySet<string> RatePairKeys(FixedFloatCurrencyResolution resolution)
    {
        var keys = new HashSet<string>(StringComparer.Ordinal);
        foreach (var currency in resolution.PayIn.Values)
        {
            keys.Add(FixedFloatRates.PairKey(currency.Code, resolution.Lightning.Code));
        }
        return keys;
    }

    private static List<FixedFloatCurrency> ReadCurrencies(JsonNode? data)
    {
        var record = FixedFloatFields.ObjectOrEmpty(data);
        var items = data as JsonArray
                    ?? record["ccies"] as JsonArray
                    ?? record["currencies"] as JsonArray
                    ?? new JsonArray();
        var currencies = new List<FixedFloatCurrency>();
        foreach (var item in items)
        {
            var entry = FixedFloatFields.ObjectOrEmpty(item);
            var code = FixedFloatFields.OptionalStringField(entry, "code") ?? FixedFloatFields.OptionalStringField(entry, "ticker");
            var coin = FixedFloatFields.OptionalStringField(entry, "coin")
                       ?? FixedFloatFields.OptionalStringField(entry, "currency")
                       ?? FixedFloatFields.OptionalStringField(entry, "symbol");
            var network = FixedFloatFields.OptionalStringField(entry, "network")
                          ?? FixedFloatFields.OptionalStringField(entry, "chain")
                          ?? FixedFloatFields.OptionalStringField(entry, "networkName")
                          ?? FixedFloatFields.OptionalStringField(entry, "name");
            if (code is not null && coin is not null && network is not null)
            {
                currencies.Add(new FixedFloatCurrency(
                    code,
                    coin.ToUpperInvariant(),
                    network,
                    FixedFloatFields.OptionalBoolean(entry, "recv"),
                    FixedFloatFields.OptionalBoolean(entry, "send")));
            }
        }
        return currencies;
    }
}
